### ISSUE TITLE
custom padding for maskable icons

### ISSUE THREAD

**Issue by glennhartmann** [w3c/manifest#805]
This is a follow-up to [Issue 555](https://github.com/w3c/manifest/issues/555).

I have Android in mind, in particular, but this may apply equally to other platforms.

[This comment](https://github.com/w3c/manifest/issues/555#issuecomment-404097653) by @mgiuca  discusses how a user agent may work around the fact that Android's adaptive icons specify a different safe zone than the web spec ([link](https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive)).

The chosen path forward, referred to as "Option B" suggests that the user agent insert a 15% padding border around the icon ([example with black padding](https://user-images.githubusercontent.com/228433/42558175-c9b31ade-8533-11e8-87b0-14982ab7a31c.png)).

The discussion assumes that the extra border will be far enough away from the centre of the icon that a user will probably never see it. Unfortunately, in my experience, this is not the case - particularly if the user is using an alternative launcher or non-circular icons. Seeing black when dragging the icon doesn't look great.

It's also suggested in [Issue 555](https://github.com/w3c/manifest/issues/555) that instead of black icons, the user agent may choose use the theme colour, average icon colour, most common icon colour, etc. This would help, but possibly still not look that great.

I'd like to consider two improvements.

**1 (short term):** Specify in the spec exactly what the border colour should be, if one is needed. Developers should know what to expect. What the colour we specify actually ends up being is less important, I think the main thing is that it's predictable and consistent across user agents. Theme colour or average icon colour both seem reasonable.

**2 (longer term):** Explore ways that we can expose custom padding to the developer. If we choose to use the theme colour, then it is already customizable to an extend. Maybe it should be a separate manifest field independent of theme colour? Maybe even optionally something more sophisticated like a larger image to composite on top of? Just spit-balling here, but a solid colour might not always be desirable.

**Comment by mgiuca** [comment]
Glenn internally shared a video which I'm waiting for permission to post here, since it's illustrative of the issue on Android.

I think it's fairly minor. The issue is specific to the fact that on Android, when you drag an icon, there is a "parallax" effect in which the icon is translated relative to its containing mask, allowing the black borders to be seen. This is only seen while dragging the icon (which is even more minor when you consider that your finger will typically be covering the entire icon during the drag).

We deliberately made the maskable icon in the manifest spec be a smaller portion of the icon than what is required on Android, because it seemed inappropriate to ask developers to supply a ridiculously large amount of padding in their icon, just to satisfy Android, when that padding would almost never be seen. This was the "Option A" in [my July 2018 post](https://github.com/w3c/manifest/issues/555#issuecomment-404097653). One important upside of Option B is that the maskable icon looks reasonable if simply displayed to the user with no modification in a square tile.

One thing to consider: is the Chrome WebAPK implementation placing the maskable icon in the _foreground_ or _background_ image that Android needs? It's supposed to be placed in the background, but perhaps it's being put in the foreground. My understanding was that the background image does not parallax when dragged on Android, only the foreground image is translated relative to the mask. So if placed in the background, you should never see the black area:

@mgiuca, [2018-07-11](https://github.com/w3c/manifest/issues/555#issuecomment-404097653):
> Now if we give this to Android, it will correctly chop down to 61% or 66%, giving the correct tight icon shown above. However, the extended border may be visible to the user in certain extreme cases. I've never seen Android show any part of the _background_ outside of the 72dp zone due to parallax, but who knows what a future version of Android might do. I think it's fairly safe to assume the user will never see these "black" pixels.

If Android actually is parallaxing the background image as well, or if the "wiggling" of the image as its dragged around is desirable, then the user agent can do more advanced padding without requiring spec intervention. My suggestions, from worst to best:

1. Fixed colour (e.g., black) padding
2. Average icon colour
3. The theme colour
4. Average or most common colour of the edge and corner pixels of the icon
5. Extrude the edges of the image out, as shown below

Average icon colour is unlikely to blend correctly with the edges, since if your icon is "white on a blue background" then the average colour will be a lighter blue than your background. Theme colour is somewhat likely to be correct, but may be completely wrong if the theme colour is your foreground colour. So option 4 is likely the best "low effort" solution.

Option 5 is likely to be the best all around, so I'll go into some (painstakingly manually done in GIMP) detail: start with a hypothetical maskable icon with non-uniform edge colour (based on [this](http://materialdesignblog.com/10-inspiring-examples-of-material-design-product-icons/)):

![image](https://user-images.githubusercontent.com/228433/66179278-a398b180-e6ab-11e9-8b10-93668428fbd6.png)

set every pixel of the 15% padding to the same colour as the nearest pixel in the source image (i.e., extrude the edges outwards in all four directions):

![image](https://user-images.githubusercontent.com/228433/66179541-bd86c400-e6ac-11e9-885e-9744d2e959db.png)

This looks weird when you look at the whole thing, but will almost certainly look "fine" if you only see a little bit of the edge while dragging. Almost all adaptive icons will feature a solid colour all the way around the edge, in which case Option 5 will be the same as Option 4 and look perfect. But for adaptive icons that have "stuff" at the edges, Option 5 will produce the best result.

> **1 (short term):** Specify in the spec exactly what the border colour should be, if one is needed.

I don't think this needs to be specified. It's a user-agent specific choice of how to present the icon, just like how we don't specify what mask to use, or what background colour to draw normal (transparent) icons on top of. We don't need to specify things that are outside of the developer's control, as long as we establish a contract between the developer and user agent, and the user agent makes a best effort attempt to fulfill that contract.

> **2 (longer term):** Explore ways that we can expose custom padding to the developer. ... Maybe it should be a separate manifest field independent of theme colour?

I don't really want to go down this path because we're effectively specifying a new field that's only useful on Android, and only useful in the edge case where the user is dragging the icon around and looking underneath their own finger to see it. I think the above options 4 and 5 should be sufficient to display the icon reasonably while parallaxing, that we shouldn't need to add more options for developers to get wrong in the manifest.

**Comment by NotWoods** [comment]
I agree that the spec is better with generalized images rather than having a special edge-case for Android. For what it's worth, Firefox is currently using options 2 and 3 but I'd like to move to option 5 in the future.

[Nova Launcher](https://medium.com/@teslacoilapps/reshaping-icons-f835f7c46e12) has some pretty interesting reshaping algorithms, but as far as I know its closed source.

**Comment by mgiuca** [comment]
> special edge-case for Android

Lol, "edge case". I see what you did there.

> Firefox is currently using options 2 and 3

2 and 3? Are you doing one of those and falling back to the other? I'd recommend switching to 4, because as I said above, in the very common case of an icon with a solid background colour all around the edge, 2 and 3 are unlikely to pick that same background colour to extend, whereas 4 (and 5) will.

**Comment by NotWoods** [comment]
At the moment we use theme color and fallback to the [Palette API](https://developer.android.com/training/material/palette-colors). I'm planning to switch to option 5 down the line, but since the edge isn't shown much (if at all) it's low priority.

**Comment by glennhartmann** [comment]
> One thing to consider: is the Chrome WebAPK implementation placing the maskable icon in the foreground or background image that Android needs? It's supposed to be placed in the background, but perhaps it's being put in the foreground. My understanding was that the background image does not parallax when dragged on Android, only the foreground image is translated relative to the mask. So if placed in the background, you should never see the black area:

Good point. Up until now, I have been placing the maskable icon in the foreground. I tried moving the maskable icon into the background instead, with a transparent foreground. This appears to solve the issue on Pixel Launcher, but it looks like Nova Launcher moves both layers at different rates, so it is still possible to see the outer border on the background image.

According to Android documentation (https://developer.android.com/guide/practices/ui_guidelines/icon_design_adaptive), "Note: Animated visual effects are generated by supported launchers. Visual effects may vary from one launcher to another." So it sounds like it's left pretty open for alternative launchers are to do whatever they want.

**Comment by glennhartmann** [comment]
> Edit: I am happy to add a (non-normative) note to the spec explaining the best practices for adapting maskable icons to platforms that expect more padding than the spec allows for, once we figure out what those best practices should be.

I'd be happy with this

**Comment by mgiuca** [comment]
> This appears to solve the issue on Pixel Launcher, but it looks like Nova Launcher moves both layers at different rates, so it is still possible to see the outer border on the background image.

Ah OK, good to know.

I still think the edge fix will be sufficient for this now-fairly-obscure case. Assigning myself to add a non-normative note to the spec about best practices for UAs.
